Follow `path` attribute on inline `mod` blocks

danjl1100 opened the pull request:
Following discussion in #335,

### Progress
- [x] Add tests based on [the reference](https://doc.rust-lang.org/reference/items/modules.html#the-path-attribute)
- [x] Add logic in `visit.rs` to follow path attributes `#[path="..."]` on modules
- [x] Verify paths do not contain `..` or **start with** `/`  (not a security boundary, but nice to have)
- [x] Report file/line/col of the `mod` statement on error

---
## Future tasks
This is a list of some enhancements that are not planned to be included in this PR:
- Add the command-line option to fail on "referent of mod not found".
    - I did verify tests pass with that `warn!` substituted with a `panic!`   (which means the `Ok(None)` return path may not be currently exercised by the test suite... it's fun thinking about how `cargo-mutants` can help improve its own test suite :slightly_smiling_face: )
- Address the TODO comment to refactor special handling of `mod.rs` to `SourceFile` itself (though the new tests should help with verifying that effort) https://github.com/sourcefrog/cargo-mutants/blob/961f1a3bc03e9d29e564f6c6b70f002c34e8437f/src/visit.rs#L469-L471

danjl1100 commented:
I started drafting the error for invalid filesystem traversal (https://github.com/sourcefrog/cargo-mutants/pull/339/commits/6c2620a74e3084b660054bd59216060eea48ad09).  It seems a bit clunky to have a "did it fail yet" `bool`; not sure how best to break out of the `Visitor` without having a flag or similar.

Do you think I should add a testdata case for this?  (named something like `dubious_paths`)
So far, I've tested it manually on a file like this.   Thinking if this is added as a test-case, I'll keep the paths harmless (replaced my test `/etc/passwd` with `/root/file`, just in case!)
```
#[path = "/root/file"]
mod slash;

#[path = "foo/../../../../root/file"]
mod dots;

#[path = "/../../../../root/file"]
mod slash_dots;
```
--
Up next I'm also planning to work on the file/line/col reporting for the current "referent of mod not found" warning.

sourcefrog commented:
Nice, thanks!

> Following discussion in #335,
> 
> ### Progress
> * [x]  Add tests based on [the reference](https://doc.rust-lang.org/reference/items/modules.html#the-path-attribute)
> * [x]  Add logic in `visit.rs` to follow path attributes `#[path="..."]` on modules
> * [ ]  Verify paths do not contain `..` or **start with** `/`  (not a security boundary, but nice to have)
> * [ ]  Report file/line/col of the `mod` statement on error
> 
> ## Future tasks
> This is a list of some enhancements that are not planned to be included in this PR:
> 
> * Add the command-line option to fail on "referent of mod not found".

I wonder if this should just be a simple error without an option to make it a warning: it should only be hit if the tree is invalid or if there's a mutants bug causing it to read the tree differently to how Cargo will. (Such as #335 for example.)

On the other hand when we do have those bugs perhaps it's nice for people still to be able to get partial results...

>   * I did verify tests pass with that `warn!` substituted with a `panic!`   (which means the `Ok(None)` return path may not be currently exercised by the test suite... it's fun thinking about how `cargo-mutants` can help improve its own test suite 🙂 )

Yep, we could add a test tree that has a dangling `mod` statement and expect that it fails.

I'm a bit surprised that this is not already caught by mutating `Ok(None)` to `Err(anyhow!(...))`.

> * Address the TODO comment to refactor special handling of `mod.rs` to `SourceFile` itself (though the new tests should help with verifying that effort) https://github.com/sourcefrog/cargo-mutants/blob/961f1a3bc03e9d29e564f6c6b70f002c34e8437f/src/visit.rs#L469-L471

sourcefrog reviewed (changes requested):
Looks good!

danjl1100 commented on src/visit.rs:
```diff
@@ -367,7 +402,41 @@ impl<'ast> Visit<'ast> for DiscoveryVisitor<'_> {
         }
-        self.mod_namespace_stack.push(mod_name.clone());
+
+        // Extract path attribute value, if any (e.g. `#[path="..."]`)
```
Good point,  I moved this to a private function and added some unit tests.

danjl1100 commented:
Thanks for the review! I updated from the discussion.

Also:
- added `testdata/dangling_mod` to verify the visitor attempts to gracefully continue on a missing module.  This seems to fix verifying the `Ok(None)` return value.
- added line/column to the warning when a module file is not found  (renamed displayed variable name from `parent_path` to `definition_site`), with an integration test to capture the output formatting

sourcefrog commented on src/visit.rs:
```diff
@@ -649,4 +762,74 @@ mod test {
+            #[path = "contains_.._dots.rs"]
+            mod dots;
+        });
+        assert_eq!(dots, Err("contains_.._dots.rs".to_owned()));
```
Actually, thinking about this more, rustc does accept path attributes that have multiple components or that contain `..`, so we should too.

```rust
#[path = "../c/c.rs"]
mod c;

fn a() {}
```

Sorry for the detour, can you pull that Err case out?

danjl1100 replied:
Sure thing. Do we still want to check the depth, to make sure we're staying inside the cargo workspace?
I'd have to check whether cargo allows relative paths outside the workspace... but it seems like cargo-mutants shouldn't try to apply mutants outside the workspace

sourcefrog replied:
Good point. It looks like rustc will follow absolute paths that point outside of the workspace, but we probably should not. Maybe we should check this at the point that we go to open the new file, making sure that it's inside the overall source directory.

danjl1100 replied:
I added a test in error_value to check for the relative path warning.  The "testdata" file copy code has some duplication, since it needs to maintain the relative paths between two testdata folders.... let me know what you think.

danjl1100 replied:
Also removed the `failed` flag as part of the rollback; it makes sense for discovery to attempt to continue if at all possible.

sourcefrog commented on src/visit.rs:
```diff
@@ -367,7 +404,23 @@ impl<'ast> Visit<'ast> for DiscoveryVisitor<'_> {
+        let mod_namespace = ModNamespace {
+            name: mod_name,
+            path_attribute,
+            source_location: node.mod_token.span().start().into(),
```
I think this will highlight specifically the `mod` token but we probably want to save the span of the `node`, i.e. the whole statement. It's not a huge deal because the keyword will obviously be close and this is relatively unlikely to be hit, but it probably sets a better example for adding more error location info in future.

danjl1100 replied:
Good point, updated to span of the whole node

sourcefrog commented:
Looks good, I really appreciate your work on this! I will make a release with it shortly.

danjl1100 commented:
Thank you for helping me along the way! 🎉

sourcefrog commented:
I'll fix the test flakes and then merge this

On Mon, Apr 29, 2024, 4:54 AM danjl1100 ***@***.***> wrote:

> Thank you for helping me along the way! 🎉
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sourcefrog/cargo-mutants/pull/339#issuecomment-2082529214>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACUR45A6U7WCWFAF4SCNHDY7YYJBAVCNFSM6AAAAABGMLZEMCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOBSGUZDSMRRGQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>
